src/core/utils/parse_cmdline.cpp

```diff
@@ -275,69 +275,86 @@ Error ParseAsIp6Prefix(const char *aString, otIp6Prefix &aPrefix)
 }
 #endif // #if OPENTHREAD_FTD || OPENTHREAD_MTD
 
-Error ParseAsHexString(const char *aString, uint8_t *aBuffer, uint16_t aSize)
+enum HexStringParseMode
 {
-    Error    error;
-    uint16_t readSize = aSize;
-
-    SuccessOrExit(error = ParseAsHexString(aString, readSize, aBuffer, kDisallowTruncate));
-    VerifyOrExit(readSize == aSize, error = kErrorInvalidArgs);
-
-exit:
-    return error;
-}
+    kModeExtactSize,   // Parse hex string expecting an exact size (number of bytes when parsed).
+    kModeUpToSize,     // Parse hex string expecting less than or equal a given size.
+    kModeAllowPartial, // Allow parsing of partial segments.
+};
 
-Error ParseAsHexString(const char *aString, uint16_t &aSize, uint8_t *aBuffer, HexStringParseMode aMode)
+static Error ParseHexString(const char *&aString, uint16_t &aSize, uint8_t *aBuffer, HexStringParseMode aMode)
 {
-    Error       error     = kErrorNone;
-    uint8_t     byte      = 0;
-    uint16_t    readBytes = 0;
-    const char *hex       = aString;
-    size_t      hexLength = strlen(aString);
-    uint8_t     numChars;
-
-    if (aMode == kDisallowTruncate)
+    Error   error        = kErrorNone;
+    size_t  stringLength = strlen(aString);
+    size_t  expectedSize = (stringLength + 1) / 2;
+    size_t  parsedSize   = 0;
+    uint8_t byte         = 0;
+    uint8_t digit;
+
+    switch (aMode)
     {
-        VerifyOrExit((hexLength + 1) / 2 <= aSize, error = kErrorInvalidArgs);
+    case kModeExtactSize:
+        VerifyOrExit(expectedSize == aSize, error = kErrorInvalidArgs);
+        break;
+    case kModeUpToSize:
+        VerifyOrExit(expectedSize <= aSize, error = kErrorInvalidArgs);
+        break;
+    case kModeAllowPartial:
+        break;
     }
 
-    // Handle the case where number of chars in hex string is odd.
-    numChars = hexLength & 1;
+    if ((stringLength & 1) != 0)
+    {
+        // If number of chars in hex string is odd, we jump to `middle`
+        // label in the loop to skip parsing the first digit.
+
+        goto middle;
```

Comment:
Can we avoid the `goto`? :)

Reply:
Changed to use a boolean flag `skipFirstDigit`.